### PR TITLE
SWS-164 Wire Graph to backend.

### PR DESCRIPTION
**Pull request by pilhuhn**
Wire the graph screen with the backend API.

Also uses breadthfirst layout so that the thingies overlap less

bildschirmfoto 2018-02-28 um 09 33 57
https://issues.jboss.org/browse/SWS-164

### CODE REVIEW

File: src/pages/ServiceGraph/ServiceGraphPage.tsx [hunk 1]
```diff
@@ -25,7 +25,16 @@ class ServiceGraphPage extends React.Component<ServiceGraphProps, ServiceGraphSt
   }
 
   componentDidMount() {
-    this.setState({ elements: FakeData.getElements() });
+    API.GetGraphElements('istio-system', null) // TODO put real namespace
```
Comment (abonas) [inline src/pages/ServiceGraph/ServiceGraphPage.tsx]: is this todo documented somewhere outside of code? JIRA?

**Comment by pilhuhn** [comment]
On 28 Feb 2018, at 12:12, abonas wrote:

> is this todo documented somewhere outside of code? JIRA?

Yes inside SWS-164 Jira and I think I also added it to SWS-165,
but can make it more obvious.

File: src/services/Api.ts [hunk 2]
```diff
@@ -31,3 +31,7 @@ const newRequest = (method: string, url: string, queryParams: any, data: any) =>
 export const GetServiceMetrics = (namespace: String, service: String, params: any) => {
   return newRequest('get', `/api/namespaces/${namespace}/services/${service}/metrics`, params, {});
 };
+
+export const GetGraphElements = (namespace: String, params: any) => {
+  return newRequest('get', `/api/namespaces/${namespace}/graphs`, params, {});
```
Comment (abonas) [inline src/services/Api.ts]: I have some concerns about this kind of REST api. What I mean - graphs is not a resource on a namespace. Users calling that kind of api do not receive any resource, they receive a specific representation of the entities/resources in that namespace. that should be an input parameter on a GET request on a namespace resource, not a separate api.

**Comment by pilhuhn** [comment]
> calling that kind of api do not receive any resource, they receive a 
> specific representation of the entities/resources in that namespace. 
> that should be an input parameter on a GET request on a namespace 
> resource, not a separate api.

Right. With that argumentation, it should perhaps even a
media type and not a queryParam that tells the type of result.

On the other hand, a resource does not necessarily be a 'leaf'
in a tree, but can be a compound (which then probably has links
to more data). E.g. a bill, that then also links out to individual line
items (descriptions).

It is exposed like this on the server for now, so the UI has to follow 
it.

**Comment by abonas** [comment]
> Right. With that argumentation, it should perhaps even a
> media type and not a queryParam that tells the type of result.
> 
> On the other hand, a resource does not necessarily be a 'leaf'
> in a tree, but can be a compound (which then probably has links
> to more data). E.g. a bill, that then also links out to individual line
> items (descriptions).
> 
> It is exposed like this on the server for now, so the UI has to follow
> it.

yes, I commented on that api on server side as well in the PRs that added it.
media type also makes sense.
having "graphs" in plural is even more confusing

**Comment by mtho11** [comment]
@pilhuhn if you merge https://github.com/swift-sunshine/swsui/pull/56 you'll get namespace param

File: src/services/Api.ts (hunk 2 above)
Comment (mwringe) [reply src/services/Api.ts]: Can you please open a JIRA for this so that we can keep track and discuss it there?
Its outside of the scope for this PR.

**Comment by mwringe** [comment]
We are missing live updating of this data, right?

**Comment by pilhuhn** [comment]
On 28 Feb 2018, at 16:52, Matthew Wringe wrote:

> We are missing live updating of this data, right?

Yes. And without this PR we miss any real data :-)

**Comment by mwringe** [comment]
True, lets skip live data and get this PR in :)

**Comment by jshaughn** [comment]
> yes, I commented on that api on server side as well in the PRs that added it. media type also makes sense. having "graphs" in plural is even more confusing

@abonas  Please add a jira for saturn team and we'll iterate on the graphs rest api in sprint 2.

File: src/pages/ServiceGraph/ServiceGraphPage.tsx (hunk 1 above)
Comment (mwringe) [reply src/pages/ServiceGraph/ServiceGraphPage.tsx]: Tracking in https://issues.jboss.org/browse/SWS-176

**Comment by mwringe** [comment]
For linking the live data to a particular namespace, that is now being tracked in https://issues.jboss.org/browse/SWS-176

**Comment by pilhuhn** [comment]
@mwringe I am waiting on #56 being merged and will re-do on top of this then.

**Comment by abonas** [comment]
@jshaughn @mwringe https://issues.jboss.org/browse/SWS-181

**Pull request merged**